BraveShared/CertificatePinning.swift

```diff
@@ -0,0 +1,160 @@
+    private let hosts: [String]
+    private let certificates: [SecCertificate]
+    private let options: PinningOptions
+    
+    public init(hosts: [String], options: PinningOptions = [.default, .validateHost]) {
+        self.hosts = hosts
+        self.options = options
+        
+        // Load certificates in the main bundle..
+        self.certificates = {
+            let paths = Set([".cer", ".CER", ".crt", ".CRT", ".der", ".DER"].map {
+                Bundle.main.paths(forResourcesOfType: $0, inDirectory: nil)
+            }.joined())
+            
+            return paths.compactMap({ path -> SecCertificate? in
+                guard let certificateData = try? Data(contentsOf: URL(fileURLWithPath: path)) as CFData else {
+                    return nil
+                }
+                return SecCertificateCreateWithData(nil, certificateData)
+            })
+        }()
+    }
+    
+    public init(hosts: [String: SecCertificate], options: PinningOptions = [.default, .validateHost]) {
+        self.hosts = hosts.map({ $0.key })
+        self.certificates = hosts.map({ $0.value })
+        self.options = options
+    }
+    
+    public func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        
+        // Certificate pinning
+        if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust {
+            if let serverTrust = challenge.protectionSpace.serverTrust {
+                do {
+                    let host = challenge.protectionSpace.host
+                    if !canPinHost(host) {
+                        throw error(reason: "Host not specified for pinning: \(host)")
+                    }
+                    
+                    try evaluate(serverTrust, forHost: host)
+                    return completionHandler(.useCredential, URLCredential(trust: serverTrust))
+                } catch {
+                    log.error(error)
+                    return completionHandler(.cancelAuthenticationChallenge, nil)
+                }
+            }
+            return completionHandler(.cancelAuthenticationChallenge, nil)
+        }
+        return completionHandler(.performDefaultHandling, nil)
+    }
+    
+    private func canPinHost(_ host: String) -> Bool {
+        return hosts.contains(host)
+    }
+    
+    private func error(reason: String) -> NSError {
+        return NSError(domain: "com.brave.pinning-certificate-evaluator", code: -1, userInfo: [NSLocalizedDescriptionKey: reason])
+    }
+    
+    public func evaluate(_ trust: SecTrust, forHost host: String) throws {
+        // Certificate validation
+        guard !certificates.isEmpty else {
+            throw error(reason: "Empty Certificates")
+        }
+        
+        // Self signed anchoring
+        if options.contains(.allowSelfSigned) {
```

Comment:
Is this just for testing? If so can we gate this behind a flag to avoid using it in actual code?

Reply:
It is for testing. We shouldn't really be using `.allowSelfSigned` at all in any sort of production code. For that reason, it is NOT part of the default options. You'd have to explicitly set this flag in the options and I don't think anyone here would ever do that. Plus during a sec-review, if we see that option set, we should automatically reject the PR :)